corehq/apps/enterprise/api/resources.py

```diff
@@ -138,12 +144,64 @@ def convert_datetime(cls, datetime_string):
         return time.isoformat()
 
 
-class ODataEnterpriseResource(ODataResource):
+class ODataEnterpriseReportResource(ODataResource):
+    REPORT_SLUG = None  # Override with correct slug
+    # If this delay is too quick, clients like PowerBI
+    # will hit their maximum number of retries before the report is ever generated.
+    # If the delay is too long, then even reports that would generate in well under the retry
+    # window will be subject to this delay (PowerBI will subject them to it twice,
+    # as the data preview and actual request perform separate queries
+    RETRY_IN_PROGRESS_DELAY = 60
+    RETRY_CONFLICT_DELAY = 120
+
     class Meta(ODataResource.Meta):
         authentication = EnterpriseODataAuthentication()
 
+    def get_object_list(self, request):
+        query_id = request.GET.get('query_id', None)
+        progress = ReportTaskProgress(
+            self.REPORT_SLUG, request.couch_user.username, query_id=query_id)
+        status = progress.get_status()
+        if status == ReportTaskProgress.STATUS_COMPLETE:
+            try:
+                # ensure this is for the same parameters
+                if progress.get_task() != self.get_report_task(request):
+                    raise ImmediateHttpResponse(
+                        response=http.HttpTooManyRequests(headers={'Retry-After': self.RETRY_CONFLICT_DELAY}))
+                data = progress.get_data()
```

Comment:
I'm finding it difficult to reason about how `TaskProgress` works in various scenarios with `query_id` passed to its constructor. We know that if `query_id` has a truthy value, then the data must exist or the request is invalid. What do you think of pulling `query_id` from the request into this method?
```suggestion
        progress = ReportTaskProgress(self.REPORT_SLUG, request.couch_user.username)
        status = ReportTaskProgress.STATUS_COMPLETE if query_id else progress.get_status()
        if status == ReportTaskProgress.STATUS_COMPLETE:
            # ensure this is for the same parameters
            if progress.get_task() != self.get_report_task(request):
                raise ImmediateHttpResponse(
                    response=http.HttpTooManyRequests(headers={'Retry-After': self.RETRY_CONFLICT_DELAY}))
            if not query_id:
                query_id = progress.get_query_id()
            try:
                data = progress.get_data(query_id)
```
`TaskProgress` can then be refactored to not have a `_query_id` attribute, which I think would make the code much easier to follow.

Reply:
This was the way I originally wrote the code, but I wanted to contain the complexity of tracking progress behind the `ReportTaskProgress` object, rather than spreading those details into this function. Specifically, it feels odd to me that the resource would re-use `ReportTaskProgress`'s `STATUS_COMPLETE` constant. Maybe there's another way to make this more readable

Reply:
I've tried to clarify this functionality in 5f8575b671887d183c1917f76db7a41c5fb9196c. I still feel that managing a task's lifecycle is best handled by a single class, rather than split between `TaskProgress` and `ODataEnterpriseReportResource`. I think the biggest source of confusion is likely that the name `query_id` doesn't exactly match what we typically think of as a query ID, so I've tried to clarify what the class is doing with a docstring.